File: sources/Android/android-communications/library/src/sdk/java/com/polar/sdk/impl/utils/PolarWatchFaceUtils.kt
// Copyright © 2026 Polar Electro Oy. All rights reserved.
package com.polar.sdk.impl.utils

import com.google.flatbuffers.FlatBufferBuilder
import com.polar.androidcommunications.api.ble.BleDeviceListener
import com.polar.androidcommunications.api.ble.BleLogger
import com.polar.androidcommunications.api.ble.model.gatt.client.psftp.BlePsFtpClient
import com.polar.androidcommunications.api.ble.model.gatt.client.psftp.BlePsFtpUtils
import com.polar.sdk.api.errors.PolarServiceNotAvailable
import com.polar.sdk.api.model.PolarWatchFaceComplication
import protocol.PftpRequest
import java.io.ByteArrayInputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder

private const val TAG = "PolarWatchFaceUtils"

data class WatchfaceConfigFields(
    val timeStyleId: Int = 0,
    val complicationLayoutId: Int = 0,
    val backgroundStyleId: Int = 0,
    val accentColor: Long = 0L,
    val complicationIds: List<Int> = emptyList(),
    val fontfaceId: Int = 0
)

internal object PolarWatchFaceUtils {

    /** KVS key for "ui.watchface_config" */
    const val WATCH_FACE_CONFIG_KVS_KEY: Int = 1064434511

    private const val FB_FIELD_TIME_STYLE_ID = 0
    private const val FB_FIELD_COMPLICATION_LAYOUT_ID = 1
    private const val FB_FIELD_BACKGROUND_STYLE_ID = 2
    private const val FB_FIELD_ACCENT_COLOR = 3
    private const val FB_FIELD_COMPLICATION_IDS = 4
    private const val FB_FIELD_FONTFACE_ID = 5
    private const val FB_TABLE_FIELD_COUNT = 6

    private const val KVTX_FILE_PATH = "/SYS/KVTX"

    /**
     * Build a KVTXScript (WRITE_BYTES + COMMIT) carrying a full WatchfaceConfig FlatBuffer.
     * Delegates generic framing to [KvtxScriptUtils.buildWriteAndCommit].
     */
    fun buildKvtxScript(fields: WatchfaceConfigFields): ByteArray {
        val configData = buildWatchFaceConfigFlatBuffer(fields)
        return KvtxScriptUtils.buildWriteAndCommit(WATCH_FACE_CONFIG_KVS_KEY, configData)
    }

    /**
     * Build a complete WatchfaceConfig FlatBuffer preserving ALL fields from [fields].
     *
     * FlatBuffers encodes scalar fields only when they differ from their default (0),
     * so unchanged zero fields are omitted automatically.
     */
    internal fun buildWatchFaceConfigFlatBuffer(fields: WatchfaceConfigFields): ByteArray {
        val builder = FlatBufferBuilder(256)

        // Build complication_ids vector first (offsets must be created before startTable)
        builder.startVector(4, fields.complicationIds.size, 4)
        for (i in fields.complicationIds.indices.reversed()) {
            builder.addInt(fields.complicationIds[i])
        }
        val vectorOffset = builder.endVector()

        builder.startTable(FB_TABLE_FIELD_COUNT)

        // field 0: time_style_id (uint16)
        builder.addShort(FB_FIELD_TIME_STYLE_ID, fields.timeStyleId.toShort(), 0)
        // field 1: complication_layout_id (uint16)
        builder.addShort(FB_FIELD_COMPLICATION_LAYOUT_ID, fields.complicationLayoutId.toShort(), 0)
        // field 2: background_style_id (uint16)
        builder.addShort(FB_FIELD_BACKGROUND_STYLE_ID, fields.backgroundStyleId.toShort(), 0)
        // field 3: accent_color (uint32) — stored as int in FlatBuffers
        builder.addInt(FB_FIELD_ACCENT_COLOR, fields.accentColor.toInt(), 0)
        // field 4: complication_ids ([int32]) — offset field
        builder.addOffset(FB_FIELD_COMPLICATION_IDS, vectorOffset, 0)
        // field 5: fontface_id (byte)
        builder.addByte(FB_FIELD_FONTFACE_ID, fields.fontfaceId.toByte(), 0)

        val tableOffset = builder.endTable()
        builder.finish(tableOffset)
        return builder.sizedByteArray()
    }

    fun extractWatchFaceConfigFromKvtxScript(script: ByteArray): ByteArray? =
        KvtxScriptUtils.extractValueForKey(script, WATCH_FACE_CONFIG_KVS_KEY)

    fun parseWatchFaceConfigFlatBuffer(raw: ByteArray): WatchfaceConfigFields {
        val empty = WatchfaceConfigFields()

        if (raw.size < 4) {
            BleLogger.w(TAG, "parseWatchFaceConfigFlatBuffer: too short (${raw.size} bytes), returning defaults")
            return empty
        }

        val bb = ByteBuffer.wrap(raw).order(ByteOrder.LITTLE_ENDIAN)
        val rootOffset = bb.getInt(0)

        if (rootOffset + 4 > raw.size) {
            BleLogger.w(TAG, "parseWatchFaceConfigFlatBuffer: rootOffset out of bounds"); return empty
        }

        val vtableOffsetFromTable = bb.getInt(rootOffset)
        val vtablePos = rootOffset - vtableOffsetFromTable

        if (vtablePos < 0 || vtablePos + 4 > raw.size) {
            BleLogger.w(TAG, "parseWatchFaceConfigFlatBuffer: vtablePos out of bounds"); return empty
        }

        val vtableSize = bb.getShort(vtablePos).toInt() and 0xFFFF
        val fieldCount = (vtableSize - 4) / 2

        // Helper: read a scalar field offset from vtable (0 = absent/default)
        fun fieldOffset(fieldIdx: Int): Int {
            if (fieldIdx >= fieldCount) return 0
            return bb.getShort(vtablePos + 4 + fieldIdx * 2).toInt() and 0xFFFF
        }

        // field 0: time_style_id (uint16)
        val timeStyleId = fieldOffset(0).let { fo ->
            if (fo == 0) 0 else (bb.getShort(rootOffset + fo).toInt() and 0xFFFF)
        }

        // field 1: complication_layout_id (uint16)
        val complicationLayoutId = fieldOffset(1).let { fo ->
            if (fo == 0) 0 else (bb.getShort(rootOffset + fo).toInt() and 0xFFFF)
        }

        // field 2: background_style_id (uint16)
        val backgroundStyleId = fieldOffset(2).let { fo ->
            if (fo == 0) 0 else (bb.getShort(rootOffset + fo).toInt() and 0xFFFF)
        }

        // field 3: accent_color (uint32)
        val accentColor = fieldOffset(3).let { fo ->
            if (fo == 0) 0L else (bb.getInt(rootOffset + fo).toLong() and 0xFFFFFFFFL)
        }

        // field 4: complication_ids ([int32])
        val complicationIds: List<Int> = run {
            val fo = fieldOffset(4)
            if (fo == 0) return@run emptyList()
            val vectorRefPos = rootOffset + fo
            if (vectorRefPos + 4 > raw.size) {
                BleLogger.w(TAG, "parseWatchFaceConfigFlatBuffer: complication_ids ref out of bounds"); return@run emptyList()
            }
            val vectorPos = vectorRefPos + bb.getInt(vectorRefPos)
            if (vectorPos + 4 > raw.size) {
                BleLogger.w(TAG, "parseWatchFaceConfigFlatBuffer: complication_ids vector out of bounds"); return@run emptyList()
            }
            val vectorLength = bb.getInt(vectorPos)
            if (vectorLength !in 0..1000) {
                BleLogger.w(TAG, "parseWatchFaceConfigFlatBuffer: complication_ids length $vectorLength invalid"); return@run emptyList()
            }
            val dataStart = vectorPos + 4
            if (dataStart + vectorLength * 4 > raw.size) {
                BleLogger.w(TAG, "parseWatchFaceConfigFlatBuffer: complication_ids data overruns buffer"); return@run emptyList()
            }
            val ids = mutableListOf<Int>()
            for (i in 0 until vectorLength) {
                val id = bb.getInt(dataStart + i * 4)
                val known = PolarWatchFaceComplication.fromId(id)
                BleLogger.d(TAG, "  complication_ids[$i] = $id (0x${id.toUInt().toString(16).padStart(8,'0')})" +
                        if (known != null) " => ${known.name}" else " => UNKNOWN (not in enum)")
                ids += id
            }
            ids
        }

        // field 5: fontface_id (byte)
        val fontfaceId = fieldOffset(5).let { fo ->
            if (fo == 0) 0 else (bb.get(rootOffset + fo).toInt() and 0xFF)
        }

        return WatchfaceConfigFields(
            timeStyleId = timeStyleId,
            complicationLayoutId = complicationLayoutId,
            backgroundStyleId = backgroundStyleId,
            accentColor = accentColor,
            complicationIds = complicationIds,
            fontfaceId = fontfaceId
        )
    }

    internal suspend fun readWatchFaceConfigFields(
        identifier: String,
        listener: BleDeviceListener?,
        handleError: (Throwable) -> Exception
    ): WatchfaceConfigFields {
        val session = PolarServiceClientUtils.sessionPsFtpClientReady(identifier, listener)
        val client = session.fetchClient(BlePsFtpUtils.RFC77_PFTP_SERVICE) as BlePsFtpClient?
            ?: throw PolarServiceNotAvailable()
        BleLogger.d(TAG, "readWatchFaceConfigFields: GET /SYS/KVTX")
        val requestBuilder = PftpRequest.PbPFtpOperation.newBuilder().apply {
            command = PftpRequest.PbPFtpOperation.Command.GET
            path = KVTX_FILE_PATH
        }
        val kvtxScript = try {
            client.request(requestBuilder.build().toByteArray()).toByteArray()
        } catch (throwable: Throwable) {
            BleLogger.e(TAG, "readWatchFaceConfigFields: GET failed: ${throwable.message}")
            throw handleError(throwable)
        }
        BleLogger.d(TAG, "readWatchFaceConfigFields: received ${kvtxScript.size} bytes")
        val flatBufferBytes = extractWatchFaceConfigFromKvtxScript(kvtxScript)
        if (flatBufferBytes == null) {
            BleLogger.w(TAG, "readWatchFaceConfigFields: key not found, returning all-defaults")
            return WatchfaceConfigFields()
        }
        return parseWatchFaceConfigFlatBuffer(flatBufferBytes)
    }

    internal suspend fun writeWatchFaceComplicationInts(
        identifier: String,
        int32Ids: List<Int>,
        listener: BleDeviceListener?,
        handleError: (Throwable) -> Exception
    ) {
        BleLogger.d(TAG, "writeWatchFaceComplicationInts: reading current config before write")
        val existingFields = readWatchFaceConfigFields(identifier, listener, handleError)
        BleLogger.d(TAG, "writeWatchFaceComplicationInts: existing=$existingFields")

        val mergedFields = existingFields.copy(complicationIds = int32Ids)
        BleLogger.d(TAG, "writeWatchFaceComplicationInts: merged=$mergedFields")

        val session = PolarServiceClientUtils.sessionPsFtpClientReady(identifier, listener)
        val client = session.fetchClient(BlePsFtpUtils.RFC77_PFTP_SERVICE) as BlePsFtpClient?
            ?: throw PolarServiceNotAvailable()
        val kvtxScript = buildKvtxScript(mergedFields)
        BleLogger.d(TAG, "writeWatchFaceComplicationInts: PUT ${kvtxScript.size} bytes to $KVTX_FILE_PATH")
        val builder = PftpRequest.PbPFtpOperation.newBuilder().apply {
            command = PftpRequest.PbPFtpOperation.Command.PUT
            path = KVTX_FILE_PATH
        }
        client.write(builder.build().toByteArray(), ByteArrayInputStream(kvtxScript)).collect {}
    }
}

File: sources/iOS/ios-communications/Sources/PolarBleSdk/sdk/impl/utils/PolarWatchFaceUtils.swift
// Copyright 2026 Polar Electro Oy. All rights reserved.

import Foundation

private let TAG = "PolarWatchFaceUtils"

// MARK: - Data model

/// Full set of watch face config fields, preserving all values across a read-modify-write cycle.
struct WatchfaceConfigFields {
    var timeStyleId: UInt16 = 0
    var complicationLayoutId: UInt16 = 0
    var backgroundStyleId: UInt16 = 0
    var accentColor: UInt32 = 0
    var complicationIds: [Int32] = []
    var fontfaceId: UInt8 = 0
}

// MARK: - Watch face utilities

internal enum PolarWatchFaceUtils {

    /// KVS key for "ui.watchface_config"
    static let WATCH_FACE_CONFIG_KVS_KEY: UInt32 = 1064434511

    private static let KVTX_FILE_PATH = "/SYS/KVTX"

    // FlatBuffer field indices
    private static let FB_FIELD_TIME_STYLE_ID:          Int = 0
    private static let FB_FIELD_COMPLICATION_LAYOUT_ID: Int = 1
    private static let FB_FIELD_BACKGROUND_STYLE_ID:    Int = 2
    private static let FB_FIELD_ACCENT_COLOR:           Int = 3
    private static let FB_FIELD_COMPLICATION_IDS:       Int = 4
    private static let FB_FIELD_FONTFACE_ID:            Int = 5
    private static let FB_TABLE_FIELD_COUNT:            Int = 6

    // MARK: - Public API

    /// Build a KVTXScript (WRITE_BYTES + COMMIT) carrying a full WatchfaceConfig FlatBuffer.
    static func buildKvtxScript(fields: WatchfaceConfigFields) -> [UInt8] {
        let configData = buildWatchFaceConfigFlatBuffer(fields: fields)
        return KvtxScriptUtils.buildWriteAndCommit(kvKey: WATCH_FACE_CONFIG_KVS_KEY, data: configData)
    }

    static func extractWatchFaceConfigFromKvtxScript(script: [UInt8]) -> [UInt8]? {
        return KvtxScriptUtils.extractValueForKey(script: script, kvKey: WATCH_FACE_CONFIG_KVS_KEY)
    }

    // MARK: - FlatBuffer encode

    /// Build a WatchfaceConfig FlatBuffer preserving all fields from `fields`.
    static func buildWatchFaceConfigFlatBuffer(fields: WatchfaceConfigFields) -> [UInt8] {
        let builder = FlatBufferBuilder(initialSize: 256)

        // Build complication_ids vector first
        builder.startVector(elemSize: 4, count: fields.complicationIds.count, alignment: 4)
        for i in stride(from: fields.complicationIds.count - 1, through: 0, by: -1) {
            builder.addRawInt32(fields.complicationIds[i])
        }
        let vectorOffset = builder.endVector(count: fields.complicationIds.count)

        builder.startTable(numFields: FB_TABLE_FIELD_COUNT)

        // field 0: time_style_id (uint16)
        builder.addInt16(field: FB_FIELD_TIME_STYLE_ID,
                         value: Int16(bitPattern: fields.timeStyleId),
                         defaultValue: 0)
        // field 1: complication_layout_id (uint16)
        builder.addInt16(field: FB_FIELD_COMPLICATION_LAYOUT_ID,
                         value: Int16(bitPattern: fields.complicationLayoutId),
                         defaultValue: 0)
        // field 2: background_style_id (uint16)
        builder.addInt16(field: FB_FIELD_BACKGROUND_STYLE_ID,
                         value: Int16(bitPattern: fields.backgroundStyleId),
                         defaultValue: 0)
        // field 3: accent_color (uint32 stored as int32)
        builder.addInt32(field: FB_FIELD_ACCENT_COLOR,
                         value: Int32(bitPattern: fields.accentColor),
                         defaultValue: 0)
        // field 4: complication_ids (vector offset)
        builder.addOffset(field: FB_FIELD_COMPLICATION_IDS,
                          value: vectorOffset,
                          defaultValue: 0)
        // field 5: fontface_id (byte)
        builder.addByte(field: FB_FIELD_FONTFACE_ID,
                        value: Int8(bitPattern: fields.fontfaceId),
                        defaultValue: 0)

        let tableOffset = builder.endTable()
        builder.finish(rootTable: tableOffset)
        return builder.sizedByteArray()
    }

    // MARK: - FlatBuffer decode

    static func parseWatchFaceConfigFlatBuffer(raw: [UInt8]) -> WatchfaceConfigFields {
        let empty = WatchfaceConfigFields()
        guard raw.count >= 4 else {
            BleLogger.trace("\(TAG): parseWatchFaceConfigFlatBuffer: too short (\(raw.count) bytes), returning defaults")
            return empty
        }

        func u16At(_ p: Int) -> Int {
            guard p + 2 <= raw.count else { return 0 }
            return Int(raw[p]) | (Int(raw[p+1]) << 8)
        }
        func i16At(_ p: Int) -> Int16 {
            return Int16(bitPattern: UInt16(u16At(p)))
        }
        func u32At(_ p: Int) -> UInt32 {
            guard p + 4 <= raw.count else { return 0 }
            return UInt32(raw[p]) | (UInt32(raw[p+1]) << 8) |
                   (UInt32(raw[p+2]) << 16) | (UInt32(raw[p+3]) << 24)
        }
        func i32At(_ p: Int) -> Int32 {
            return Int32(bitPattern: u32At(p))
        }

        let rootOffset = Int(u32At(0))
        guard rootOffset + 4 <= raw.count else {
            BleLogger.trace("\(TAG): parseWatchFaceConfigFlatBuffer: rootOffset out of bounds")
            return empty
        }

        let vtableOffsetFromTable = Int(i32At(rootOffset))
        let vtablePos = rootOffset - vtableOffsetFromTable
        guard vtablePos >= 0, vtablePos + 4 <= raw.count else {
            BleLogger.trace("\(TAG): parseWatchFaceConfigFlatBuffer: vtablePos out of bounds")
            return empty
        }

        let vtableSize = u16At(vtablePos)
        let fieldCount = (vtableSize - 4) / 2

        func fieldOffset(_ fieldIdx: Int) -> Int {
            guard fieldIdx < fieldCount else { return 0 }
            return u16At(vtablePos + 4 + fieldIdx * 2)
        }

        // field 0: time_style_id (uint16)
        let timeStyleId: UInt16 = {
            let fo = fieldOffset(0); guard fo != 0 else { return 0 }
            return UInt16(bitPattern: i16At(rootOffset + fo))
        }()

        // field 1: complication_layout_id (uint16)
        let complicationLayoutId: UInt16 = {
            let fo = fieldOffset(1); guard fo != 0 else { return 0 }
            return UInt16(bitPattern: i16At(rootOffset + fo))
        }()

        // field 2: background_style_id (uint16)
        let backgroundStyleId: UInt16 = {
            let fo = fieldOffset(2); guard fo != 0 else { return 0 }
            return UInt16(bitPattern: i16At(rootOffset + fo))
        }()

        // field 3: accent_color (uint32)
        let accentColor: UInt32 = {
            let fo = fieldOffset(3); guard fo != 0 else { return 0 }
            return u32At(rootOffset + fo)
        }()

        // field 4: complication_ids ([int32])
        let complicationIds: [Int32] = {
            let fo = fieldOffset(4)
            guard fo != 0 else { return [] }
            let vectorRefPos = rootOffset + fo
            guard vectorRefPos + 4 <= raw.count else {
                BleLogger.trace("\(TAG): parseWatchFaceConfigFlatBuffer: complication_ids ref out of bounds")
                return []
            }
            let vectorPos = vectorRefPos + Int(i32At(vectorRefPos))
            guard vectorPos + 4 <= raw.count else {
                BleLogger.trace("\(TAG): parseWatchFaceConfigFlatBuffer: complication_ids vector out of bounds")
                return []
            }
            let vectorLength = Int(i32At(vectorPos))
            guard vectorLength >= 0, vectorLength <= 1000 else {
                BleLogger.trace("\(TAG): parseWatchFaceConfigFlatBuffer: complication_ids length \(vectorLength) invalid")
                return []
            }
            let dataStart = vectorPos + 4
            guard dataStart + vectorLength * 4 <= raw.count else {
                BleLogger.trace("\(TAG): parseWatchFaceConfigFlatBuffer: complication_ids data overruns buffer")
                return []
            }
            var ids = [Int32]()
            for i in 0..<vectorLength {
                let id = i32At(dataStart + i * 4)
                let known = PolarWatchFaceComplication.fromId(id)
                BleLogger.trace("\(TAG): complication_ids[\(i)] = \(id) (0x\(String(UInt32(bitPattern: id), radix: 16)))" +
                    (known != nil ? " => \(known!)" : " => UNKNOWN"))
                ids.append(id)
            }
            return ids
        }()

        // field 5: fontface_id (byte)
        let fontfaceId: UInt8 = {
            let fo = fieldOffset(5); guard fo != 0 else { return 0 }
            return raw[rootOffset + fo]
        }()

        return WatchfaceConfigFields(
            timeStyleId: timeStyleId,
            complicationLayoutId: complicationLayoutId,
            backgroundStyleId: backgroundStyleId,
            accentColor: accentColor,
            complicationIds: complicationIds,
            fontfaceId: fontfaceId
        )
    }

    // MARK: - PFTP read / write

    static func readWatchFaceConfigFields(_ identifier: String,
                                          serviceClientUtils: PolarServiceClientUtils) async throws -> WatchfaceConfigFields {
        let session = try serviceClientUtils.sessionFtpClientReady(identifier)
        guard let client = session.fetchGattClient(BlePsFtpClient.PSFTP_SERVICE) as? BlePsFtpClient else {
            throw PolarErrors.serviceNotFound
        }
        var operation = Protocol_PbPFtpOperation()
        operation.command = .get
        operation.path = KVTX_FILE_PATH
        BleLogger.trace("\(TAG): readWatchFaceConfigFields: GET \(KVTX_FILE_PATH)")
        let requestData = try operation.serializedData()
        let responseData = try await client.request(requestData)
        let bytes = [UInt8](responseData)
        BleLogger.trace("\(TAG): readWatchFaceConfigFields: received \(bytes.count) bytes")
        if let flatBufferBytes = extractWatchFaceConfigFromKvtxScript(script: bytes) {
            return parseWatchFaceConfigFlatBuffer(raw: flatBufferBytes)
        } else {
            BleLogger.trace("\(TAG): readWatchFaceConfigFields: key not found, returning defaults")
            return WatchfaceConfigFields()
        }
    }

    static func writeWatchFaceComplicationInts(_ identifier: String,
                                               int32Ids: [Int32],
                                               serviceClientUtils: PolarServiceClientUtils) async throws {
        var existingFields = try await readWatchFaceConfigFields(identifier, serviceClientUtils: serviceClientUtils)
        existingFields.complicationIds = int32Ids
        BleLogger.trace("\(TAG): writeWatchFaceComplicationInts: merged=\(existingFields)")

        let session = try serviceClientUtils.sessionFtpClientReady(identifier)
        guard let client = session.fetchGattClient(BlePsFtpClient.PSFTP_SERVICE) as? BlePsFtpClient else {
            throw PolarErrors.serviceNotFound
        }
        let kvtxScript = buildKvtxScript(fields: existingFields)
        BleLogger.trace("\(TAG): writeWatchFaceComplicationInts: PUT \(kvtxScript.count) bytes to \(KVTX_FILE_PATH)")

        var operation = Protocol_PbPFtpOperation()
        operation.command = .put
        operation.path = KVTX_FILE_PATH
        let proto = try operation.serializedData()
        let inputStream = InputStream(data: Data(kvtxScript))
        for try await _ in client.write(proto as NSData, data: inputStream) {}
    }
}
